Framework/API/inc/MantidAPI/WorkspacePropertyWithIndex.tcc

```diff
@@ -170,21 +173,39 @@ const SpectrumIndexSet WorkspacePropertyWithIndex<TYPE>::getIndices() const {
   if (list.size() == 0)
     return indexInfo.makeIndexSet();
 
-  switch (m_indexTypeProp->selectedType()) {
-  case IndexType::SpectrumNumber:
-    return indexInfo.makeIndexSet(std::vector<Mantid::Indexing::SpectrumNumber>(
-        list.begin(), list.end()));
-    break;
-  case IndexType::WorkspaceIndex:
-    return indexInfo.makeIndexSet(
-        std::vector<Mantid::Indexing::GlobalSpectrumIndex>(list.begin(),
-                                                           list.end()));
-    break;
+  // Determine if input is a pure range or just a list of numbers
+  auto res = std::minmax_element(list.begin(), list.end());
+  auto minIndex = res.first - list.begin();
+  auto maxIndex = res.second - list.begin();
+  bool isPureRange = (list[maxIndex] - list[minIndex] + 1) == list.size();
```

Comment:
1. If spectrum 3 does not exist, an exception will be thrown which will be caught in the validator. The user would have entered invalid input. Not sure what the problem is.
2. The problem I think is that ArrayProperty may be inefficient for long ranges. Might be an idea to augment ArrayProperty to handle pure ranges more effectively? Also, how should we handle something like `1:33,45:55,76,78` which I think is perfectly valid input.